File: c/binary_tree.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct node
{
    int value;
    struct node *left;
    struct node *right;
} node;

typedef struct BinaryTree
{
    node *root;
} BinaryTree;

void insert(BinaryTree *tree, int value)
{
    node *new_node = (node *)malloc(sizeof(node));
    new_node->left = NULL;
    new_node->right = NULL;
    new_node->value = value;
    node *current_node = tree->root;
    if (current_node == NULL)
    {
        tree->root = new_node;
        return;
    }
    while (current_node != NULL)
    {
        if (current_node->value > value)
        {
            if (current_node->left == NULL)
            {
                current_node->left = new_node;
                return;
            }
            current_node = current_node->left;
        }
        else
        {
            if (current_node->right == NULL)
            {
                current_node->right = new_node;
                return;
            }
            current_node = current_node->right;
        }
    }
}

bool lookup(BinaryTree *tree, int value)
{
    node *current_node = tree->root;
    while (current_node != NULL)
    {
        if (current_node->value > value)
            current_node = current_node->left;
        else if (current_node->value < value)
            current_node = current_node->right;
        else
            return true;
    }
    return false;
}

node *delete_node(node *Node, int value)
{
    if (Node == NULL)
        return NULL;
    if (Node->value == value)
    {
        if (Node->left == NULL && Node->right == NULL)
            return NULL;
        if (Node->left == NULL)
            return Node->right;
        if (Node->right == NULL)
            return Node->left;
        node *tmp = Node->right;
        while (tmp->left)
            tmp = tmp->left;
        Node->value = tmp->value;
        Node->right = delete_node(Node->right, tmp->value);
        tmp->left = Node->left;
return Node->right;
    }
    else if (Node->value > value)
        Node->left = delete_node(Node->left, value);
    else
        Node->right = delete_node(Node->right, value);
    return Node;
}

void delete(BinaryTree *tree, int value)
{
    tree->root = delete_node(tree->root, value);
}

void printTreeHelper(node *node, int space) {
    if (node == NULL)
        return;
    int i;
    printTreeHelper(node->right, space+5);
    printf("\n");
    for (i = 0; i < space; i++) {
        printf(" ");
    }
    printf("%d\n", node->value);
    printTreeHelper(node->left, space+5);
}

void printTree(node *root) {
    printTreeHelper(root, 0);
}

int main(void)
{
    node *root = (node *)malloc(sizeof(node));
    root->value = 15;
    root->left = NULL;
    root->right = NULL;
    BinaryTree *tree = (BinaryTree *)malloc(sizeof(BinaryTree));
    tree->root = root;
    insert(tree, 8);
    insert(tree, 5);
    insert(tree, 2);
    insert(tree, 6);
    insert(tree, 10);
    insert(tree, 9);
    insert(tree, 11);
    printTree(tree->root);
}

File: python/binary_tree.py

class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        if self.root == None:
            self.root = Node(value)
            return 
        root = self.root
        while (root != None):
            if root.value > value:
                if root.left == None:
                    root.left = Node(value)
                    return 
                root = root.left
            else:
                if root.right == None:
                    root.right = Node(value)
                    return
                root = root.right
        
    def lookup(self, value):
        root = self.root
        while (root != None):
            if root.value > value:
                root = root.left
            elif root.value < value:
                root = root.right
            else:
                return True
        return False
    
    # def remove(self, root, value):
    #     if not root:
    #         return False
        
    #     if value < root.value:
    #         root = self.remove(root.left, value)
    #     elif value > root.value:
    #         root = self.remove(root.right, value)
    #     else:
    #         if not root.left:
    #             return root.right
    #         elif not root.right:
    #             return root.left
    #         else:
    #             next = root.right
    #             while next.left:
    #                 next = next.left
    #             root.value = next.value
    #             self.remove(root.right, next.value)
    #     return root
    
    def remove(self, value):
        def remove_node(node, value):
            if not node:
                return None
            if node.value == value:
                if not node.left and not node.right:
                    return None
                if not node.left:
                    return node.right
                if not node.right:
                    return node.left
                temp = node.right
                while temp.left:
                    temp = temp.left
                node.value = temp.value
                node.right = remove_node(node.right, temp.value)
            elif node.value > value:
                node.left = remove_node(node.left, value)
            else:
                node.right = remove_node(node.right, value)
            return node

        self.root = remove_node(self.root, value)
            
    def print_tree(self):
        if self.root == None:
            print("Empty binary search tree")
            return
        self.print_node(self.root, "", True)

    def print_node(self, node, prefix, is_left):
        if node == None:
            return
        print(f"{prefix}{'|-- ' if is_left else '`-- '}{node.value}")
        self.print_node(node.left, prefix + ("|   " if is_left else "    "), True)
        self.print_node(node.right, prefix + ("|   " if not is_left else "    "), False)



if __name__ == "__main__":
    Tree = BinaryTree()
    Tree.insert(8)
    Tree.insert(5)
    Tree.insert(2)
    Tree.insert(6)
    Tree.insert(10)
    Tree.insert(9)
    Tree.print_tree()
    Tree.remove(8)
    Tree.print_tree()
